lib/Infer/AbstractInterpreter.cpp

```diff
@@ -200,185 +217,236 @@ namespace souper {
-      if (getSMin(lhs).sge(getSMax(rhs)))
-	Result.Zero.setBit(0);
+      if (getSMax(LHS).slt(getSMin(RHS)))
+        Result.One.setBit(0);
+      if (getSMin(LHS).sge(getSMax(RHS)))
+        Result.Zero.setBit(0);
       return Result;
     }
 
-    llvm::KnownBits ule(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+    llvm::KnownBits ule(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS) {
       llvm::KnownBits Result(1);
-      if (getUMax(lhs).ule(getUMin(rhs)))
-	Result.One.setBit(0);
-      if (getUMin(lhs).ugt(getUMax(rhs)))
-	Result.Zero.setBit(0);
+      if (getUMax(LHS).ule(getUMin(RHS)))
+        Result.One.setBit(0);
+      if (getUMin(LHS).ugt(getUMax(RHS)))
+        Result.Zero.setBit(0);
       return Result;
     }
 
-    llvm::KnownBits sle(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+    llvm::KnownBits sle(const llvm::KnownBits &LHS, const llvm::KnownBits &RHS) {
       llvm::KnownBits Result(1);
-      if (getSMax(lhs).sle(getSMin(rhs)))
-	Result.One.setBit(0);
-      if (getSMin(lhs).sgt(getSMax(rhs)))
-	Result.Zero.setBit(0);
+      if (getSMax(LHS).sle(getSMin(RHS)))
+        Result.One.setBit(0);
+      if (getSMin(LHS).sgt(getSMax(RHS)))
+        Result.Zero.setBit(0);
       return Result;
     }
   }
 
-  bool isReservedConst(Inst *I) {
-    return I->K == Inst::ReservedConst ||
-          (I->K == Inst::Var &&
-          (I->Name.find(ReservedConstPrefix) != std::string::npos));
-  }
-
-  bool isReservedInst(Inst *I) {
-    return I->K == Inst::ReservedInst ||
-          (I->K == Inst::Var &&
-          (I->Name.find(ReservedInstPrefix) != std::string::npos));
-  }
-
   bool isConcrete(Inst *I, bool ConsiderConsts, bool ConsiderHoles) {
     return !hasGivenInst(I, [ConsiderConsts, ConsiderHoles](Inst* instr) {
-			      if (ConsiderConsts && isReservedConst(instr))
-				return true;
-			      if (ConsiderHoles && isReservedInst(instr))
-				return true;
-
-			      return false;
-			    });
+      if (ConsiderConsts && isReservedConst(instr))
+        return true;
+      if (ConsiderHoles && isReservedInst(instr))
+        return true;
+      return false;
+    });
   }
 
-  EvalValue getValue(Inst *I, ValueCache &C, bool PartialEval) {
+  // Tries to get the concrete value from @I. If PartialEval is enabled, it tries to
+  // launch concrete interpreter to get the value. If not possible, it returns
+  // EvalValue of ValueKind::Unimplemented
+  EvalValue getValue(Inst *I, ConcreteInterpreter &CI, bool PartialEval) {
     if (I->K == Inst::Const)
       return {I->Val};
-    if (C.find(I) != C.end()) {
-      return C[I];
-    } else {
-      if (PartialEval && isConcrete(I)) {
-        return evaluateInst(I, C);
-      }
-    }
+    else if (I->K == Inst::Var && !isReservedConst(I))
+      // evaluateInst will only give us input value of the variable; it doesn't
+      // evaluate anything.
+      return CI.evaluateInst(I);
+
+    if (PartialEval && isConcrete(I))
```

Comment:
yeah, I can't think of a good reason not to just do partial evaluation whenever it's possible

Reply:
@manasij7479 ^

Reply:
The unit tests rely on this feature, iirc.

Reply:
they should be changed